finding_resources.md

```diff
@@ -0,0 +1,303 @@
+# Finding Resources
+
+This document offers guidelines and rationales for the various mechanisms
+used throughout Ignition to find resources (files) at runtime. File types
+include, but are not limited to, plugins (shared libraries), world
+descriptions (SDF files) and graphics files like meshes and textures.
+
+This is not meant to list all the specifics of how Ignition finds each resource,
+although it may mention some as examples. For an exhaustive explanation of how
+each type of resource is found, please check
+[Ignition Gazebo's Finding resources tutorial](https://ignitionrobotics.org/api/gazebo/3.0/resources.html).
+
+**Table of Contents**
+
+* [Precedence](#precedence)
+* [Ways of finding resources](#ways-of-finding-resources)
+   * [Full path](#full-path)
+   * [Full URL](#full-url)
+   * [Custom callbacks](#custom-callbacks)
+   * [Relative paths](#relative-paths)
+      * [Schemes](#schemes)
+      * [Prefixes](#prefixes)
+
+## Precedence
+
+There are many ways of referring to a file. Likewise, one path may match
+several files.
+
+For example, let's assume that all these files exist:
+
+```
+/tmp/worlds/shapes.sdf
+/tmp/more_worlds/shapes.sdf
+/home/user/shapes.sdf
+/usr/share/ignition/ignition-gazebo2/worlds/shapes.sdf
+```
+
+When the user does the following, which file gets loaded?
+
+```
+cd /home/user
+IGN_GAZEBO_RESOURCE_PATH=/tmp/worlds:/more_worlds
+ign gazebo shapes.sdf
+```
+
+It's necessary to have a consistent way of prioritizing which approach
+precedes others, so that the user knows what to expect.
+
+The general guideline for precedence is to go **from the most custom to the most
+standard**. The most custom methods can be easily changed by the user at runtime,
+while the least custom ones would require recompiling a library.
+
+On the example above, it's most likely safe to assume that the user
+didn't want to load the world that is installed with Ignition Gazebo, otherwise
+they wouldn't have gone through the trouble of moving directory and setting
+an environment variable. But in the end, we can never know for sure what the
+user really meant. They could happen to be in that directory, and
+the variable may have been set by another application that appended it to
+`~/.bashrc`.
+
+That is to say that there's no perfect solution. But at least by keeping
+consistency users can refer back to these guidelines and understand
+what to expect.
+
+Thus, Ignition should look for resources in the following order:
+
+1. Full path
+2. Environment variables
+3. Path relative to current directory
+4. Common home directory
+5. Custom callbacks set by downstream developers
```

Comment:
Downstream might be a bit confusing here because `ign-gazebo` sets custom callbacks, I think, and I'm not sure a reader of this document would consider `ign-gazebo` to be downstream.

Reply:
Yeah it's tough to write the document in a way general enough that it applies to all libraries. Technically, `ign-gazebo` is downstream from `ign-gui`, for example. I was hoping to capture that, as well as 3rd-party libraries. Hopefully changing "downstream developers" to "downstream projects" makes things clearer?